ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -80,66 +86,97 @@ public void run() {
             timer.schedule(refreshIngestionResourceValuesTask, 0, REFRESH_INGESTION_RESOURCES_PERIOD);
 
         } catch (Exception e) {
-            log.error(String.format("Error in initializing ResourceManager: %s.", e.getMessage()), e);
+            log.error("Error in initializing ResourceManager: {}.", e.getMessage(), e);
             throw e;
         }
     }
 
-    public void clean() {
+    void clean() {
         ingestionResources.clear();
     }
 
-    public String getKustoIdentityToken() throws Exception {
-        if (identityToken == null) {
-            refreshIngestionAuthToken();
-            if (identityToken == null) {
-                throw new Exception("Unable to get Identity token");
+    String getIngestionResource(ResourceTypes resourceType) throws Exception {
+        if (!ingestionResources.containsKey(resourceType)) {
+            // When the value is not available, we need to get the tokens from Kusto (refresh):
+            refreshIngestionResources();
+            try{
+                // If the write lock is locked, then the read will wait here.
+                // In other words if the refresh is running yet, then wait until it ends:
+                ingestionResourcesLock.readLock().lock();
+                if (!ingestionResources.containsKey(resourceType)) {
+                    throw new Exception("Unable to get ingestion resources for this type: " + resourceType.getName());
+                }
+            } finally {
+                ingestionResourcesLock.readLock().unlock();
             }
         }
-        return identityToken;
+        return ingestionResources.get(resourceType).nextValue();
```

Comment:
Shouldn't that be under the lock?

Reply:
The implementation of the refreshIngestionResources() was changed so it finishes retrieving the resource values into a new resources map, and only then it replaces the old resources map with the new resources map in one atomic operation. Therefore, when calling the getIngestionResource() method, if the key is available, it will return it instantly. Only if the key is not available it runs refresh with write lock into it, and read lock in this method to ensure returning the value when the refresh finishes.
(If during periodic refresh, this method was called, then it will return the available value before the refresh ends, and this is OK.)